### PR TITLE
ikfast_kinematics_plugin: Write XML files as UTF-8

### PR DESCRIPTION
**Pull request by mintar**
No more garbled author names for people with an umlaut in their name! :)

### CODE REVIEW

**Comment by 130s** [comment]
So is [this line](https://github.com/uos/katana_driver/commit/9eb6583b8c5a1202931b805ed7bd2cbfcd49e01a#diff-9ac9c2bb476b9bb5132f6d239835e1bdL8) of the change what you re-generated using the suggested change? Assuming so +1.

**Comment by mintar** [comment]
@130s : yes, that line and [this line](https://github.com/uos/katana_driver/commit/9eb6583b8c5a1202931b805ed7bd2cbfcd49e01a#diff-d399146e7bd315cf030bb4f9234f5050L1), and the corresponding change in `package.xml`, but I had already changed it to `UTF-8` before so it's not visible in the diff:

```diff
-<?xml version='1.0' encoding='ASCII'?>
+<?xml version='1.0' encoding='UTF-8'?>
```

**Comment by gavanderhoorn** [comment]
@mintar: I like this.

Would it be a good idea to add some tests for this?

**Comment by mintar** [comment]
> Would it be a good idea to add some tests for this?

Yes, the whole ikfast package generator could use some tests. However, this is not trivial. Perhaps this could be integrated with [moveit_kinematics_tests](https://github.com/ros-planning/moveit_kinematics_tests) somehow?

**Comment by mintar** [comment]
I've added another commit.

Also, I've found one case where this doesn't work: If neither the build nor run depends nor the exports are changed, `modified_pkg` will be false:  https://github.com/mintar/moveit/blob/fee3f60394d5e35c5c23dd64dcad9ffd93bb6b49/moveit_kinematics/ikfast_kinematics_plugin/scripts/create_ikfast_moveit_plugin.py#L309

In that case, it doesn't write the package.xml, so it doesn't change `encoding='ASCII'` to `encoding='UTF-8'`. I'm working on that.

**Comment by mintar** [comment]
AFAICT, there is no way to get the information whether the input `package.xml` was encoded as ASCII or UTF-8 from `lxml`, because the XML declaration is not part of the element tree. That leaves us with the following 3 possibilities:

1. Leave it as it is. The `package.xml` won't be regenerated if the switch to UTF-8 encoding is the only change; if there are changes to the `depends` or `export` tags, the switch to UTF-8 will also happen.
2. Open the original `package.xml` as a regular text file and check if the first line is `<?xml version='1.0' encoding='UTF-8'?>`; if it isn't, regenerate the file.
3. Always regenerate the `package.xml`. This would destroy any whitespace changes by the user; all meaningful content would be kept intact (as with the other options).

Which do you prefer? All are easy to implement. Option (3) would simplify the script a bit.

**Comment by v4hn** [comment]
> Always regenerate the package.xml. This would destroy any whitespace changes by the user; all meaningful content would be kept intact (as with the other options).

Please provide a patch for this solution. And thanks a lot also in my name. :-)

**Comment by mintar** [comment]
Ok, done and ready to merge.

File: moveit_kinematics/ikfast_kinematics_plugin/scripts/create_ikfast_moveit_plugin.py [hunk 1]
```diff
@@ -310,9 +304,10 @@ def update_deps(reqd_deps, req_type, e_parent):
 
    if not found:
       export_element.append(new_export)
-      with open(package_file_name,"w") as f:
-         package_xml.write(f, xml_declaration=True, pretty_print=True)
-      print '\nModified package.xml at \''+package_file_name+'\''
+
```
Comment (davetcoleman) [inline moveit_kinematics/ikfast_kinematics_plugin/scripts/create_ikfast_moveit_plugin.py]: Could you add a comment here, something like:

> Always write the package xml file, even if there are no changes, to ensure proper encodings are used in the future (UTF-8)

**Comment by mintar** [comment]
@davetcoleman : Done!

**Comment by v4hn** [comment]
Merged. Thanks again @mintar !

**Comment by 130s** [comment]
Should this be backported?

**Comment by mintar** [comment]
There's nothing preventing it from running on Indigo/Jade, but it's not essential to backport this IMO.

**Pull request merged**